patches/server/0007-Add-EntityTeleportHinderedEvent.patch

```diff
@@ -0,0 +1,147 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Mariell Hoversholm <proximyst@proximyst.com>
+Date: Tue, 28 Apr 2020 19:59:02 +0200
+Subject: [PATCH] Add EntityTeleportHinderedEvent
+
+
+diff --git a/src/main/java/net/minecraft/server/BlockEnderPortal.java b/src/main/java/net/minecraft/server/BlockEnderPortal.java
+index 213388c74545063cec2de51e6b80cec88d85479a..4dde23c4040b0d66ca130fd0358eff82b7ad2598 100644
+--- a/src/main/java/net/minecraft/server/BlockEnderPortal.java
++++ b/src/main/java/net/minecraft/server/BlockEnderPortal.java
+@@ -3,6 +3,8 @@ package net.minecraft.server;
+ // CraftBukkit start
+ import org.bukkit.event.entity.EntityPortalEnterEvent;
+ import org.bukkit.event.player.PlayerTeleportEvent;
++import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause; // Papercut
++import us.minevict.papercut.event.entity.EntityTeleportHinderedEvent; // Papercut
+ // CraftBukkit end
+ 
+ public class BlockEnderPortal extends BlockTileEntity {
+@@ -25,7 +27,23 @@ public class BlockEnderPortal extends BlockTileEntity {
+ 
+     @Override
+     public void a(IBlockData iblockdata, World world, BlockPosition blockposition, Entity entity) {
+-        if (!world.isClientSide && !entity.isPassenger() && !entity.isVehicle() && entity.canPortal() && VoxelShapes.c(VoxelShapes.a(entity.getBoundingBox().d((double) (-blockposition.getX()), (double) (-blockposition.getY()), (double) (-blockposition.getZ()))), iblockdata.getShape(world, blockposition), OperatorBoolean.AND)) {
++        if (!world.isClientSide && entity.canPortal() && VoxelShapes.c(VoxelShapes.a(entity.getBoundingBox().d((double) (-blockposition.getX()), (double) (-blockposition.getY()), (double) (-blockposition.getZ()))), iblockdata.getShape(world, blockposition), OperatorBoolean.AND)) { // Papercut
++            // Papercut start
++            if (entity.isPassenger()) { // Moved down from wrapping statement
++                if (new EntityTeleportHinderedEvent(entity.getBukkitEntity(), EntityTeleportHinderedEvent.Reason.IS_PASSENGER, TeleportCause.END_PORTAL).callEvent()) {
```

Comment:
Is there any reason as to why you don't combine both ifs into one here and in the statements bellow?

Reply:
The `EntityTeleportHinderedEvent.Reason` is different for every statement.

Reply:
I meant in the sense of
`if (condition && Event#callEvent() /* if it attempts to retry */)`

Reply:
Because it always has to return, but it might not always want to retry. If it retries and the condition is still false, it has to return still. There is a warning on the API that this might induce a thread-lock.

Reply:
You are RIGHT!